kiwi/lib/search.py

```diff
@@ -0,0 +1,413 @@
+            f'Setting result to -1'
+        )
+
+    logger.info(f"RESULTS: {result} {main_metric}")
+    logger.info(f"PARAMETERS: {dict(zip(names, values))}")
+    for name, value in zip(names, values):
+        logger.info(f'{name}: {value}')
+
+    return result
+
+
+def setup_run(directory: Path, seed: int, debug=False, quiet=False) -> Path:
+    if not directory.exists():
+        # Initialize a new folder with name '0'
+        output_dir = directory / '0'
+        output_dir.mkdir(parents=True)
+    else:
+        # Initialize a new folder incrementing folder count
+        output_dir = directory / str(len(list(directory.glob('*'))))
+        if output_dir.exists():
+            backup_directory = output_dir.with_name(
+                f'{output_dir.name}_{datetime.now().isoformat()}'
+            )
+            logger.warning(
+                f'Folder {output_dir} already exists; moving it to {backup_directory}'
+            )
+        output_dir.mkdir(parents=True)
+
+    logger.info(f'Initializing new search folder at: {output_dir}')
+
+    configure_logging(output_dir=output_dir, verbose=debug, quiet=quiet)
+    configure_seed(seed)
+
+    return output_dir
+
+
+def run(config: Configuration) -> Study:
+
+    output_dir = setup_run(config.output.directory, config.seed)
+
+    logger.info(f'Saving all Optuna results to: {output_dir}')
+
+    # FIXME: this is not neat (but it does work)
+    hydra._internal.hydra.GlobalHydra().clear()
+    kiwi_config = kiwi.cli.arguments_to_configuration(
+        {'CONFIG_FILE': config.base_config}
+    )
+
+    if not kiwi_config['run'].get('use_mlflow'):
+        logger.warning(
+            'You are not using MLflow logging but it is highly advised to do so; '
+            'set `run.use_mlflow=true` in the `base_config` file to enable MLflow'
+        )
+
+    kiwi_config['trainer']['checkpoint'][
+        'early_stop_patience'
+    ] = config.options.patience
+    kiwi_config['trainer']['checkpoint'][
+        'validation_steps'
+    ] = config.options.validation_steps
+
+    # Initialize or load a study
+    if config.load_study:
+        logger.info(f'Loading study to resume from {config.load_study}')
+        study = joblib.load(config.load_study)
+    else:
+        logger.info('Initializing study')
+        pruner = optuna.pruners.MedianPruner()
+        study = optuna.create_study(
+            direction='maximize', pruner=pruner, sampler=TPESampler(seed=config.seed)
+        )
+
+    # Optimize the study
+    try:
+        study.optimize(
+            partial(objective, config=config, kiwi_config=kiwi_config),
+            n_trials=config.num_trials,
+        )
+    except Exception as e:
+        logger.error(
+            f'Error occured during search: {e}; '
+            f'current best params are {study.best_params}'
+        )
+    except KeyboardInterrupt:
+        logger.info('Early stopping search caused by ctrl-C')
```

Comment:
I believe these two exceptions need to be be reversed? The first one catches all exceptions already AFAIK

Reply:
Yes, order is wrong.